src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_FACTOR = 1.5;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForIndex(index);
+        ensureCapacity();
+        for (int i = size; i > index; i--) {
+            elementData[i] = elementData[i - 1];
+        }
+        elementData[index] = value;
+        size++;
 
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null || list.isEmpty()) {
```

Comment:
The condition 'list == null || list.isEmpty()' is doing two things: checking for null and checking if the list is empty. According to the checklist, we should not complicate if-else constructions. It would be more readable to split this into two separate checks, and throw different exceptions or return early if the list is null.